### PR TITLE
Add `typescript.implementationsCodeLens.showOnInterfaceMethods` setting (#136282)

### LINKED ISSUE #136282: Show "implementations" code lens for TypeScript interface methods
The code lens feature shows both "references" and "implementations" for interfaces.
However, it only shows "references" for its methods.

It would be awesome if next to method "references" there was also "implementations".

image

This would be very useful for interfaces that have many methods and many references.

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This closes #136282

It also refreshes codelenses when the existing `typescript.referencesCodeLens.showOnAllFunctions` setting changes.

### CODE REVIEW

**Comment by gjsjohnmurray** [comment]
Test case:
```ts
interface IFoo {
    bar(): void;
}

class Foo implements IFoo {
    bar() {
        console.log('bar');
    }
}

function doSomething(): void {
    return;
}

new Foo().bar();
new Foo().bar();
new Foo().bar();
new Foo().bar();
new Foo().bar();
new Foo().bar();

doSomething();
```

Setting `"typescript.implementationsCodeLens.showOnInterfaceMethods": true` will add a `1 implementation` codelens above line 2.

Also, toggling setting `typescript.referencesCodeLens.showOnAllFunctions` will show/hide a `1 reference` codelens above line 11.

**Comment by hediet** [comment]
Oh nice, thank you!

File: extensions/typescript-language-features/src/languageFeatures/codeLens/baseCodeLensProvider.ts [hunk 1]
```diff
@@ -22,6 +22,9 @@ export class ReferencesCodeLens extends vscode.CodeLens {
 }
 
 export abstract class TypeScriptBaseCodeLensProvider implements vscode.CodeLensProvider<ReferencesCodeLens> {
+	protected readonly subscriptions: vscode.Disposable[] = [];
+	protected changeEmitter = new vscode.EventEmitter<void>();
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/codeLens/baseCodeLensProvider.ts:26]: This emitter should also get disposed of. You can inherit from `Disposable` too to make managing the disposables a little cleaner

**Review by mjbvz — CHANGES_REQUESTED** [review]
Looks great! Thanks

Just would like to see one small change to how the disposables are managed

File: extensions/typescript-language-features/src/languageFeatures/codeLens/baseCodeLensProvider.ts (hunk 1 above)
Comment (gjsjohnmurray) [reply extensions/typescript-language-features/src/languageFeatures/codeLens/baseCodeLensProvider.ts:26]: Thanks for the feedback.

> This emitter should also get disposed of. 

I derived the code from here:

https://github.com/microsoft/vscode/blob/0d552b5d26d922d68f87fde61382657fc5bc885b/extensions/npm/src/npmScriptLens.ts#L32-L35

Does that code also fail to dispose its emitter, or did I overlook how/where it does the right thing?

> You can inherit from Disposable too to make managing the disposables a little cleaner

By simply appending `, Disposable` to what `TypeScriptBaseCodeLensProvider` implements? Please help me understand what that will achieves.

Comment (mjbvz) [reply extensions/typescript-language-features/src/languageFeatures/codeLens/baseCodeLensProvider.ts:26]: Yes it's likely missing there too

If you inherit from `Disposable`, then you don't have to implement `dispose()` yourself. Instead you can just call `this._register` with every disposable value

Comment (gjsjohnmurray) [reply extensions/typescript-language-features/src/languageFeatures/codeLens/baseCodeLensProvider.ts]: I have pushed what I think you meant.

**Review by mjbvz — APPROVED** [review]

**Review by DonJayamanne — APPROVED** [review]

**Pull request merged**